### PR TITLE
introduce standalone estimate gas type

### PR DESCRIPTION
**Pull request by stevencartavia**
#12250

### CODE REVIEW

File: crates/rpc/rpc-eth-api/src/helpers/estimate.rs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+use alloy_primitives::U256;
+use alloy_rpc_types::{state::StateOverride, BlockId};
+use alloy_rpc_types_eth::transaction::TransactionRequest;
+use futures::Future;
+use std::sync::Arc;
+
+/// Gas estimator implementation
+#[derive(Debug)]
+pub struct GasEstimator<T> {
+    inner: Arc<T>,
```
Comment (mattsse) [inline crates/rpc/rpc-eth-api/src/helpers/estimate.rs:14]: we can remove the arc here

File: crates/rpc/rpc-eth-api/src/helpers/estimate.rs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        Self { inner }
+    }
+
+    /// Estimate gas needed for execution of the `request` at the [`BlockId`].
+    pub async fn estimate_gas_at(
+        &self,
+        request: TransactionRequest,
+        at: BlockId,
+        state_override: Option<StateOverride>,
+    ) -> impl Future<Output = Result<U256, T::Error>> + Send {
```
Comment (mattsse) [inline crates/rpc/rpc-eth-api/src/helpers/estimate.rs:39]: we def want this as a new trait `trait GasEstimator`

File: crates/rpc/rpc-eth-api/src/helpers/estimate.rs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+        async move {
+            let (cfg, block_env, at) = inner.evm_env_at(at).await?;
+
+            inner
+                .spawn_blocking_io(move |this| {
+                    let state = this.state_at_block_id(at)?;
+                    this.estimate_gas_with(cfg, block_env, request, state, state_override)
+                })
+                .await
```
Comment (mattsse) [inline crates/rpc/rpc-eth-api/src/helpers/estimate.rs:50]: all this does here is reuse the existing estimate trait.

what we want instead is to extract this into a standalone type

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/rpc/rpc-eth-api/src/helpers/call.rs
Comment (mattsse) [inline crates/rpc/rpc-eth-api/src/helpers/call.rs:1]: this must be updated so that the estimate functions are removed here

**Review by mattsse — CHANGES_REQUESTED** [review]

**Review by mattsse — APPROVED** [review]
I've completed this by updating some bounds, because this isn't super obvious

**Comment by mattsse** [comment]
cc @tcoratger

**Comment by tcoratger** [comment]
> cc @tcoratger
> 
> this decouples estimation but could likely need more work

@mattsse @stevencartavia thanks guys :)

**Pull request merged**